Electronic structure navigation

ndaelman-hu opened the pull request:
Aims to structure polymorphic spin sections within a larger electronic hierarchy (laid out #212).

ndaelman-hu committed:
- Introduce `BaseSpinState`, `SimpleSpinState`, `NonCollinearSpinState`
- Repurpose `OrbitalsState` to `ElectronicState`
-- TODO: Update other `OrbitalsState`
- Reference `ElectronicState` from `ModelSystem`

ndaelman-hu committed:
Remove `orbital_ref` to disambiguate `electronic_state`

ndaelman-hu committed:
- Define BaseSpinState._degeneracy and implement in subsections
- Simplify `ElectronicState.resolve_degeneracy()`
- Simplify `CoreHole` and `HubbardInteractions` methods by redefining the schema types

ndaelman-hu committed:
- Substitute `OrbitalsState` for `ElectronicState`
-- TODO: run out to other files, e.g. `ModelMethod`

ndaelman-hu committed:
correct comment

ndaelman-hu commented:
Thx to @EBB2675 for spotting that the previous branch had old commits in it. Has been resolved by cherry-picking based from `develop`.

Moreover, any failing tests or formatting is handled at the end.

JosePizarro3 commented:
Hey, I always wanted to restructure this part of the schema regarding the quantum numbers. Please, make sure the Python logic works as you intend + the theory behind is sounded (feel free to contribute here: https://github.com/nomad-coe/nomad-simulations/discussions/22)

EBB2675 commented on src/nomad_simulations/schema_packages/atoms_state.py:
```diff
@@ -3,38 +3,127 @@
+            if self.mj_quantum_number is not None:
+                # Specific mj values are defined
+                mjs = RussellSaundersState.generate_MJs(
+                    J=self.j_quantum_number[0], rising=True
```
should this be `J=jj` instead of `J=self.j_quantum_number`

ndaelman-hu committed:
Introduce `BaseSpinOrbitalState`

ndaelman-hu committed:
Add validation and normalization for `BaseSpinOrbitalState`

ndaelman-hu committed:
Strongly restructure to have spin-orbit always together. Instead, split along symmetry, e.g. spherical symmetric.
TODO: split up `SphericalSymmetryState` to simplify logic.

ndaelman-hu committed:
Merge branch 'develop' into 'add_system_orbitals'

ndaelman-hu committed:
- Clean redundant code
- Add author context to a few comments (lost during merge)
- Fix and clean imports

Changes to be committed:
	modified:   src/nomad_simulations/schema_packages/atoms_state.py

ndaelman-hu committed:
- Fix mypy errors
-- Add optional None output to get_label
-- Clean logic `resolve_occupation`

EBB2675 commented on src/nomad_simulations/schema_packages/atoms_state.py:
```diff
@@ -1,109 +1,371 @@
+# TODO: add `class CrystalFieldState(BaseSpinOrbitalState)`
+
+
+class ElectronicState(Entity):
```
Could we change the semantics here from `ElectronicState` to something like `SpinOrbitalState` (or `OneElectronState`)?

In many subfields, electronic state typically refers to a many-electron wavefunction, e.g., a configuration state function or a linear combination of Slater determinants

Here, as I understand it, the current `ElectronicState` section represents a one-electron spin-orbital definition under a given coupling regime (LS, jj, relativistic). Am i right?

To avoid confusion between single-electron and many-electron levels of description, would it be clearer to rename it to `SpinOrbitalState` or a similar term that explicitly signals the one-electron nature of this entity?

ndaelman-hu replied:
> Here, as I understand it, the current ElectronicState section represents a one-electron spin-orbital definition under a given coupling regime (LS, jj, relativistic). Am i right?

Actually, `ElectronicState` is meant to represent the multi-electronic state corresponding to a particular `ModelSystem`. This includes also `AtomsState` and 1-electron states.

But you're right, its `__init__` is too specific top single-electronic systems in a spherical potential. I'll see to move that up, to SphericalSymmetryState`.

ndaelman-hu commented:
@EBB2675  the tests are not passing, since the old infrastructure still references `OrbitalsState`. I focused instead more on drafting the new layout.

I can see to hook it up with the rest if the schema, if that helps. If we want to change design a lot, then it might be better to hold off, though. Your call.

**Edit:** just to clarify, I've already merge `develop` into this branch, as you requested (commit hash: 1614ac5bb645d538a6e2ed2f167884b1fc976940, follow-up commits are me just cleaning up a bit).

EBB2675 commented:
is the intention following?


ElectronicState               # generic (multi-electron system)
│
├── SphericalSymmetryState    # 1-electron states with V(r)
├── MolecularOrbitalState     # 1-electron states in point-group symmetry
└── BlochOrbitalState         # 1-electron states in periodic potentials

ndaelman-hu commented:
> is the intention following?
> 
> ElectronicState # generic (multi-electron system) 
> ├── SphericalSymmetryState # 1-electron states with V(r)
> ├── MolecularOrbitalState # 1-electron states in point-group symmetry
> └── BlochOrbitalState # 1-electron states in periodic potentials

Indeed, something along those lines. I'm guessing that `# 1-electron states in periodic potentials` means a list of single-electron states?

Atm, it only contains (the parent class to) `SphericalSymmetryState`, though. Similar to `ModelSystem`, it may have other nested electronic states. Lemme know if it's better to discuss in a meeting.

ndaelman-hu committed:
- Move `ElectronicState.__init_()` to `SphericalSymmetryState`
- Add provisionary quantum symbols to `SphericalSymmetryState`
- Add reference to `AtomsState` and the substate count tol `ElectronicState`

Changes to be committed:
	modified:   src/nomad_simulations/schema_packages/atoms_state.py

ndaelman-hu committed:
- Introduce `populate_atoms_state_refs` to set ElectronicState.atoms_state_ref recursively

Changes to be committed:
	modified:   src/nomad_simulations/schema_packages/atoms_state.py

ndaelman-hu committed:
Migrate quantum number validation and mapping to `SphericalSymmetryState`

ndaelman-hu committed:
- Update `OrbitalsState` dependencies to `ElectronicStates`
- Update interfaces interacting

Changes to be committed:
	modified:   src/nomad_simulations/schema_packages/model_method.py
	modified:   src/nomad_simulations/schema_packages/properties/greens_function.py
	modified:   src/nomad_simulations/schema_packages/properties/spectral_profile.py

ndaelman-hu committed:
Apply ruff

ndaelman-hu committed:
Update EelctronicState interface and descriptions to clarify usage and role

ndaelman-hu commented:
As a bit more context to changes up to commit 5e2d51e:

In this refactor, I’ve made `ElectronicState` the central hub for all electronic structure data. It now acts as the main entry point for navigating electronic properties from an input-perspective. From here, we can talk about any set of atoms having an associated quantum state and projections of their properties onto substates.

`ElectronicState` is not necessarily meant to capture the exact numerics of the linear combinations or the populations. While it contains the `basis_orbitals`, this just references the quantum state objects necessary to describe or understand the state. Further quantum numbers describing any level in `ElectronicState` can be added modularly, by inheriting from `BaseSpinOrbitalState`.

By consolidating electronic structure data in this way, we ensure a consistent and extensible interface for downstream analysis, property calculation, and workflow integration. The result is a much cleaner interface: electronic structure is always accessed via `ElectronicState`, and the details of each orbital are encapsulated in quantum state objects referenced within it.

EBB2675 commented on src/nomad_simulations/schema_packages/atoms_state.py:
```diff
@@ -1,157 +1,248 @@
+        description="""
+        Symbolic representation of the `l` quantum number, i.e., 's', 'p', 'd', 'f'.
+        """,
+    )  # @EBB2675: do we want to serialize symbolic representations? or maybe just provide a setter / getter mapping?
```
I’d avoid serializing symbolic shell labels and keep the canonical storage purely numeric:

- Store: l: int (0=s, 1=p, 2=d, 3=f, 4=g, …).

- Expose: a derived l_symbol property (getter) and a validating setter that maps symbols ↔ integers.

most codes and libraries reason about l numerically; symbols are pure presentation anyhow

ndaelman-hu replied:
Sounds good. I'd also prefer only 1 form, but supporting the other via a mapping.
My main aim is querying more so than representation. I don't think we support mapping queries in NOMAD, but would absolutely love that.

ndaelman-hu replied:
@TLCFEM @ladinesa Any ideas how do to a query over a mapping well?
In the old GUI, I guess one would write the mapping in JavaScript. Are there different strategies now?

ndaelman-hu replied:
we'll keep this open for when we write any queries

EBB2675 commented on src/nomad_simulations/schema_packages/atoms_state.py:
```diff
@@ -1,157 +1,248 @@
+
+class SphericalSymmetryState(
+    BaseSpinOrbitalState
+):  # @EBB2675 we could also split this section into 3 mutually inheriting sections
```
makes sense for the future but I’d keep this as a single section for now. Once requirements are clearer, we can add symmetry-specific sections (e.g., `CrystalFieldState`) to avoid premature complexity

ndaelman-hu replied:
Okay, let's keep it for a follow-up then. I'll slightly modify the comment to reflect this and resolve this convo

EBB2675 commented on src/nomad_simulations/schema_packages/atoms_state.py:
```diff
@@ -1,157 +1,248 @@
+
+    kappa_quantum_number = Quantity(
+        type=np.int32, description='κ = ±(j + 1/2), encodes l and j'
+    )  # ? should this be mutually exclusive with j_quantum_number?
```
I’d suggest not making them mutually exclusive. κ and (j, l) encode the same information, so it’s reasonable to accept either; if both are provided, we can check consistency (as your helpers already do) and, if helpful, derive the missing value during normalization

ndaelman-hu replied:
okay. given that kappa is quite specific to a relativistic setting, where j and l are still proper quantum numbers. So, populating j and l from kappa makes contextual sense. The other way around, I'm less sure.

EBB2675 commented on src/nomad_simulations/schema_packages/atoms_state.py:
```diff
@@ -235,87 +324,449 @@ def resolve_number_and_symbol(
 
     n_excited_electrons = Quantity(
-        type=np.float64,
+        type=unit_float,  # ? too restrictive
```
I'd keep unit_float here 👍

EBB2675 commented on src/nomad_simulations/schema_packages/atoms_state.py:
```diff
@@ -235,87 +324,449 @@ def resolve_number_and_symbol(
+    )
+
+    basis_orbitals = SubSection(
+        section_def=BaseSpinOrbitalState.m_def, # @EBB2675: do you see numerical_settings.basis_set also fit here?
```
Looks good to me, just to confirm my understanding:

we have `basis_orbitals` as semantic labels (AOs/SALCs/Wannier). We can later link to the numerical AO layer via `numerical_basis_ref` and an `ao_indices` array that points into the expanded AO list in `basis_set.AtomicOrbitals`

ndaelman-hu replied:
That's the plan! Since we already have abstract numerical classes, I could also try to add a connection now. Do you have a suggested structure?

EBB2675 commented on src/nomad_simulations/schema_packages/atoms_state.py:
```diff
@@ -235,87 +324,449 @@ def resolve_number_and_symbol(
+        total_degeneracy = orbital_degeneracy × spin_degeneracy
+
+        The orbital degeneracy is 2*l + 1 (if ml not specified) or 1 (if ml specified).
+        The spin degeneracy is delegated to the spin_state._degeneracy property, which
```
```suggestion
        The spin degeneracy is delegated to the spin_orbit_state._degeneracy property, which
```

EBB2675 commented on src/nomad_simulations/schema_packages/atoms_state.py:
```diff
@@ -235,87 +324,449 @@ def resolve_number_and_symbol(
+            2 * self.l_quantum_number + 1 if self.ml_quantum_number is None else 1
+        )
+        spin_degeneracy = (
+            self.spin_state._degeneracy if self.spin_state is not None else 2
```
```suggestion
            self.spin_orbit_state._degeneracy if self.spin_orbit_state is not None else 2
```

ndaelman-hu replied:
I think this got addressed indirectly during other follow-up commits.

EBB2675 commented on src/nomad_simulations/schema_packages/atoms_state.py:
```diff
@@ -1,157 +1,248 @@
+            kappa: Relativistic quantum number
+
+        Returns:
+            l = |κ| - 1/2 if κ < 0, or l = κ for κ > 0
```
```suggestion
            l = |κ| - 1 if κ < 0, or l = κ for κ > 0 (with κ ≠ 0)
```

EBB2675 commented on src/nomad_simulations/schema_packages/atoms_state.py:
```diff
@@ -235,87 +324,449 @@ def resolve_number_and_symbol(
         return quantity
 
-    def resolve_degeneracy(self) -> int | None:
+    def _russel_saunders_j_values(
```
```suggestion
    def _russell_saunders_j_values(
```

EBB2675 commented on src/nomad_simulations/schema_packages/atoms_state.py:
```diff
@@ -235,87 +324,449 @@ def resolve_number_and_symbol(
+        self._orbitals = {
+            -1: dict(zip(range(4), ('s', 'p', 'd', 'f'))),
+            0: {0: ''},
+            1: dict(zip(range(-1, 2), ('x', 'z', 'y'))),
```
this is a bit unconventional, some packages ( Gaussian/PySCF) use  (x, y, z), some others (Molden/CP2K) use (z, x, y) 

and tying ml = −1,0,+1 directly to real labels (px, py, pz) is itself convention-dependent


do you think it is better not to hard-code the label orders?

EBB2675 replied:
i have checked again, the most used convention is z,x,y

ndaelman-hu replied:
will fall back to getters and remove these maps

EBB2675 commented on src/nomad_simulations/schema_packages/atoms_state.py:
```diff
@@ -1,157 +1,248 @@
+
+    j_quantum_number = Quantity(
+        type=np.float64,
+        shape=['1..2'],
```
I think the intent behind `shape=['1..2']` is to capture a small manifold (e.g., j=l ± 1/2​). For an actual state, though, j should be a single value. Could we make `j_quantum_number` scalar and compute candidate j values via a helper when needed?

ndaelman-hu replied:
Agreed, this would be a good a use case of `ElectronicState` actually.

The question is: how do we deal with a system that defines only the `l_quantum_number`? It would be a system with `self._degeneracy == 2`.
So, in this case, we can just leave it be (my preference), or generate 2 subsections with the individual `j_quantum_numbers`. Any preferences?

EBB2675 replied:
i was thinking of storing one `ElectronicState` with `j_quantum_number=None` when only `l` is known.

tbh i do not have specific preferences, we can indeed leave it be and see at the heavy parsing stage 👍

ndaelman-hu replied:
> i was thinking of storing one ElectronicState with j_quantum_number=None when only l is known.

That's indeed better, and more aligned with the proposal here. Lemme see to make this extra clear.

EBB2675 reviewed (approved):
Looks great from my side, I added super minor suggestions


I think this came up before, i will open an issue (#269)  for the following:

We’re hard-coding ℓ/mℓ labels and orders (px/py/pz, dxy…) in multiple places in the code. It'd be great if we can centralize this, either in our own utilities or via something external, e.g. `libcint`

ndaelman-hu committed:
refactor: migrate from OrbitalsState to ElectronicState navigation

- Fix quantum number type definitions (strictly_positive_int, positive_int)
- Add SphericalSymmetryState._degeneracy property with j-prioritized calculation
- Implement ElectronicState.resolve_degeneracy() using spin_orbit_state._degeneracy
- Add CoreHole-specific degeneracy resolution using orbital_ref
- Update all property imports to use ElectronicState instead of OrbitalsState
- Migrate ModelSystem to use electronic_state with basis_orbitals pattern

BREAKING CHANGE: OrbitalsState removed, use ElectronicState.basis_orbitals instead

ndaelman-hu committed:
test: update test infrastructure for ElectronicState schema

- Replace OrbitalsState with ElectronicState and SphericalSymmetryState imports
- Add make_spherical_state() and make_electronic_state() helper functions
- Update generate_model_system to build electronic_state with basis_orbitals
- Update generate_simulation_electronic_dos to use ElectronicState wrappers
- Modernize type annotations and improve test utilities

ndaelman-hu committed:
test: migrate all tests to ElectronicState schema

- Rename TestOrbitalsState to TestSphericalSymmetryState
- Add _degeneracy_from_orbital helper using ElectronicState wrapper
- Update test cases for j-prioritized degeneracy calculation logic
- Fix CoreHole and HubbardInteractions test expectations
- Update model_method tests to use ElectronicState navigation
- Remove deprecated OrbitalsState references across all test files
- Ensure complete test coverage with new schema structure

ndaelman-hu committed:
Apply ruff

ndaelman-hu committed:
fix: resolve mypy type errors in model_system and numerical_settings

- Fix bond_list array shape type consistency in model_system.py
- Ensure filtered bonds maintain (N, 2) shape after masking and unique operations
- Fix return type in numerical_settings.py resolve_points_and_offset()
- Convert numpy meshgrid tuple to list for type compatibility
- Convert monkhorst_pack single array to list format as expected by signature

ndaelman-hu committed:
correct typo in Russell-Saunders

Co-authored-by: Esma B. Boydas <80156958+EBB2675@users.noreply.github.com>

ndaelman-hu committed:
Correct formula orbital quantum number in terms of relativistic quantum number

Co-authored-by: Esma B. Boydas <80156958+EBB2675@users.noreply.github.com>

ndaelman-hu committed:
Resolve reviewer feedback on j.l vs kappa

ndaelman-hu committed:
- Limit `SphericalSymmetryState.j_quantum_number` to a single value rather than a list

Changes to be committed:
	modified:   src/nomad_simulations/schema_packages/atoms_state.py
	modified:   tests/conftest.py
	modified:   tests/test_atoms_state.py

ndaelman-hu committed:
- Relegate quantum symbols back to getter symbols

Changes to be committed:
	modified:   src/nomad_simulations/schema_packages/atoms_state.py
	modified:   tests/conftest.py
	modified:   tests/test_atoms_state.py

ndaelman-hu committed:
- Add docs of how quantum numbers are related, indicating how they can be derived from each other
- Add note on m_l symbols and magnetic axis alignment
- Make j_quantum_number and mj_quantum_number scalars (type: positive_float())

Changes to be committed:
        modified:   src/nomad_simulations/schema_packages/atoms_state.py
	modified:   tests/test_atoms_state.py

ndaelman-hu committed:
- Deactivate automated kappa and j derivations, as it may give the wrong impressions.
-- Retain conversion capabilities
-- Add TODO for future integration
- Update tests to stop using quantum symbols

Changes to be committed:
	modified:   src/nomad_simulations/schema_packages/atoms_state.py
	modified:   tests/test_model_method.py

ndaelman-hu committed:
- Implement `SphericalSymmetryState._name`
-- Add test

Changes to be committed:
	modified:   src/nomad_simulations/schema_packages/atoms_state.py
	modified:   tests/test_atoms_state.py

ndaelman-hu committed:
- Correct symbol testing

Changes to be committed:
	modified:   tests/test_model_method.py

ndaelman-hu committed:
- Explain design philosophy in `docs`

Changes to be committed:
	new file:   docs/atoms_state/electronic_states.md

ndaelman-hu committed:
Merge branch 'develop' into add_system_orbitals

Resolved conflicts:
- atoms_state.py: Added Section import
- model_system.py: Added pint, Context, Section imports; kept type annotations for mypy
- numerical_settings.py: Used develop's improved array handling
- greens_function.py: Kept ElectronicState (new refactored version)
- spectral_profile.py: Added Section import

All conflicts resolved in favor of keeping the ElectronicState/SphericalSymmetryState
refactoring while incorporating develop's improvements for type safety and array handling.

ndaelman-hu committed:
Fix critical bugs in TB.resolve_orbital_references and update tests

Fix two critical bugs in model_method.py:
- Fix numpy array boolean evaluation: particle_indices=[0] was incorrectly
  treated as empty because bool(np.array([0])) returns False. Use explicit
  None check and len() comparison instead.
- Fix extend vs append: orbitals_ref.extend() was iterating over
  ElectronicState fields instead of appending the object itself.

Update tests to use ElectronicState/SphericalSymmetryState schema:
- Repurpose commented-out test cases (9) and (10) with proper structure
- Fix docstring references from OrbitalsState to SphericalSymmetryState
- Fix docstring example to use correct field name electronic_state
- Remove unused imports from test files

All 20 test cases now passing.

ndaelman-hu committed:
Refactor ElectronicState and DOSProfile architecture

This commit implements several architectural improvements to clean up
the ElectronicState and DOSProfile design:

1. Remove atoms_state_ref from ElectronicState
   - Eliminates circular reference back to AtomsState
   - Removes populate_atoms_state_refs() method
   - Relies on metainfo m_parent navigation instead

2. Merge CoreHole.orbital_ref into inherited spin_orbit_state
   - Removes redundant orbital_ref field from CoreHole
   - Uses inherited spin_orbit_state for consistency
   - Removes CoreHole.resolve_degeneracy() override

3. Add ElectronicState.get_atoms_state() helper method
   - Encapsulates m_parent navigation logic
   - Works for both top-level and sub-state ElectronicState instances
   - Provides clean API for finding parent AtomsState

4. Refactor DOSProfile to use ElectronicState as gateway
   - DOSProfile.entity_ref now always points to ElectronicState
   - Uses spin_orbit_state presence to distinguish atom vs orbital DOS
   - Atom DOS: ElectronicState with spin_orbit_state=None
   - Orbital DOS: ElectronicState with spin_orbit_state set
   - Updates resolve_pdos_name() and generate_from_projected_dos()

5. Fix ElectronicState.normalize() bug
   - Corrects n_quantum_number check to look in spin_orbit_state
   - Ensures ElectronicState.name is properly set during normalization

All 673 tests pass with these changes.

ndaelman-hu committed:
Extract ElectronicState naming logic into get_name() method

This commit refactors the name generation logic in ElectronicState:

1. Add ElectronicState.get_name() method
   - Combines spin_orbit_state name with parent AtomsState label
   - Returns 'orbital_name atom_label' for orbital-level states
   - Returns 'atom_label' for atom-level states (no spin_orbit_state)
   - Uses get_atoms_state() helper for parent navigation

2. Simplify normalize() method
   - Now just calls self.name = self.get_name()
   - Cleaner separation of concerns

3. Simplify DOSProfile.resolve_pdos_name()
   - ElectronicState.name already includes atom label
   - Just prepends 'orbital' or 'atom' prefix
   - Significantly reduces code duplication

4. Update test fixtures
   - Add explicit normalization of top-level ElectronicState
   - Ensures names are properly set before PDOS generation

All 673 tests pass with these changes.

ndaelman-hu committed:
Make get_parent_entity() generic to support ModelSystem and AtomsState

This commit refactors the parent navigation to be more generic:

1. Refactor get_parent_entity() to work with any parent type
   - Recursively navigates up through ElectronicState hierarchy
   - Stops at first non-ElectronicState parent (AtomsState or ModelSystem)
   - No longer uses isinstance checks for specific types

2. Update get_name() to handle both parent types
   - AtomsState provides get_label() returning chemical symbol (e.g., 'Ga')
   - ModelSystem provides name attribute (e.g., 'GaAs_bulk')
   - Added clarifying comments for expected behavior

3. Remove deprecated get_atoms_state() method
   - Replaced with more generic get_parent_entity()
   - Updated spectral_profile.py to use get_parent_entity()

All 673 tests pass with these changes.

ndaelman-hu committed:
Apply ruff formatting

ndaelman-hu committed:
Refactor Green's Functions to use Pattern A (ElectronicState gateway)

Apply Pattern A to BaseGreensFunction and QuasiparticleWeight classes:

1. BaseGreensFunction refactoring:
   - Remove atoms_state_ref field (deprecated Pattern B)
   - Rename correlated_orbitals_ref → entity_ref
   - Update docstring to document ElectronicState gateway pattern
   - Remove comment about enforcing order with dual references

2. QuasiparticleWeight refactoring:
   - Remove atoms_state_ref field (deprecated Pattern B)
   - Rename correlated_orbitals_ref → entity_ref
   - Update docstring to document gateway pattern
   - Update n_atoms description to note derivation from entity_ref

3. Inherited classes automatically adopt new pattern:
   - ElectronicGreensFunction
   - ElectronicSelfEnergy
   - HybridizationFunction

The ElectronicState gateway pattern allows accessing parent AtomsState
via entity_ref.get_parent_entity(), eliminating circular references
and aligning with the architecture established in DOSProfile.

All 673 tests pass with these changes.

ndaelman-hu committed:
Refactor Occupancy class to use Pattern A (ElectronicState gateway)

Complete Pattern A migration by refactoring the Occupancy class:

1. Remove atoms_state_ref field (deprecated Pattern B)
2. Update orbitals_state_ref description to document parent navigation
3. Update class docstring to explain ElectronicState gateway pattern
4. Update value description to remove atoms_state_ref reference

With this change, all property classes now consistently use Pattern A:
- Single orbital/entity reference pointing to ElectronicState
- Navigate to parent AtomsState via get_parent_entity()
- No circular references or dual reference patterns

All 673 tests pass with these changes.

ndaelman-hu commented:
@EBB2675 @JosePizarro3 I further polished the new interface and went thoroughly over its roll-out.
@ladinesa I will soon also push any downstream modifications on the parsers' side.

Feel free to review on your own initiative. Given the time pressure, and that @EBB2675 seemed content with the PR thusfar, I _would like to aim for merging by tomorrow evening_. Pls write me, in case you feel a review is urgent, but does not fall within that timeframe.

ndaelman-hu committed:
Update documentation to reflect Pattern A architecture

Update electronic_states.md to document the new Pattern A architecture:
- Reference fields now point to ElectronicState instances
- Navigation to parent AtomsState via get_parent_entity() helper
- Eliminates circular references while maintaining clean hierarchy

This completes the documentation updates for the Pattern A refactoring.

ladinesa commented on src/nomad_simulations/schema_packages/atoms_state.py:
```diff
@@ -14,145 +13,301 @@
-        """,
-    )
+    @property
+    def _name(self) -> str:
```
you can add `@abstractmethod`

ndaelman-hu replied:
done, also applied to `BaseSpinOrbitalState._degeneracy`

ladinesa commented on src/nomad_simulations/schema_packages/atoms_state.py:
```diff
@@ -14,145 +13,301 @@
-        """,
-    )
+    @property
+    def _degeneracy(self) -> int:
```
you normally use property decorator to cache the variable but it appears you did not implement it.

ndaelman-hu replied:
added caching via `@cached_property`
also rolled out to `_name`, in this class and its child classes

ladinesa commented on src/nomad_simulations/schema_packages/atoms_state.py:
```diff
@@ -14,145 +13,301 @@
 
-    ms_quantum_number = Quantity(
+    @property
+    def l_quantum_symbol(self) -> str:
```
i am really skeptical about using property decorator for MSections as it somehow confuses the `property` idea (quantities and sub-sections). I would rather have getter methods

ndaelman-hu replied:
I see what you mean. Could you elaborate a bit more on how you see `@property` fit into NOMAD?

ladinesa commented on src/nomad_simulations/schema_packages/atoms_state.py:
```diff
@@ -14,145 +13,301 @@
+                expected_j = self.compute_j_from_kappa(self.kappa_quantum_number)
+
+                if abs(self.j_quantum_number - expected_j) >= 1e-10:
+                    logger.error(
```
Avoid formatted strings in logs

ndaelman-hu replied:
applied throughout

ndaelman-hu committed:
Apply reviewer feedback to enhance code quality:

  1. Add caching to _name and _degeneracy properties
     - Convert @property to @cached_property for BaseSpinOrbitalState._name and _degeneracy
     - Convert @property to @cached_property for SphericalSymmetryState._name and _degeneracy

  2. Add @staticmethod decorator to helper methods
     - Mark compute_j_from_kappa() as static
     - Mark compute_l_from_kappa() as static

  3. Replace f-strings with %-formatting in logger calls
     - Update validate_kappa_j_relationship() error messages
     - Update validate_quantum_numbers() error messages

Changes to be committed:
	modified:   src/nomad_simulations/schema_packages/atoms_state.py

ndaelman-hu committed:
- Roll out reviewer feedback to other files too
- Apply ruff linting

Changes to be committed:
	modified:   src/nomad_simulations/schema_packages/atoms_state.py
	modified:   src/nomad_simulations/schema_packages/model_method.py
	modified:   src/nomad_simulations/schema_packages/model_system.py

ndaelman-hu committed:
Merge remote-tracking branch 'origin/develop' into add_system_orbitals

ndaelman-hu commented:
@ladinesa are you fine with me merging this? I applied all but 1 of your suggestions.

ladinesa commented:
> @ladinesa are you fine with me merging this? I applied all but 1 of your suggestions.

go ahead please
